### DISCUSSION TITLE
Offline messages lost when user refreshes the page or closes the tab

### DISCUSSION THREAD

**Discussion by akshaybande81** [processone/ejabberd#3734] [Questions] [answered]
I am using strophe.js as client for ejabberd server.

The problem I am facing is that, offline messages are lost when user refreshes the page or closes the tab.

I am using anon login. mod_offline is enabled for ejabberd server.

```yaml
auth_method:
      - odbc
      - anonymous
    allow_multiple_connections: true
    anonymous_protocol: both
```

Steps to replicate:

login to xmpp with full jid.
for ex. user@domain.com/resource_1

Send negative present to store offline messages for delayed delivery. Messages are getting stored at this step. I can confirm offline message count > 0 using ejabberd get_offline_count API

As I refresh the page or close the tab, messages are lost. I can confirm this using ejabberd get_offline_count API

If i reconnect again with same jid but different resource for ex. user@domain.com/resource_2, I won't receive any of the offline messages.

I am suspecting the ejabberd clears the offline messages for anon users. But I can't find any mention of this online, am I missing something here?

**Comment by badlop** [comment, marked as answer]
> I am suspecting the ejabberd clears the offline messages for anon users. But I can't find any mention of this online, am I missing something here?

Right, that's how it works:
1. When the client logouts, ejabberd_auth_anonymous:unregister_connection gets called
2. That function runs the "remove_user" hook
3. mod_offline:remove_user removes the offline messages of that user

If that cleaning mechanism were not implemented, the database would get polluted with useless offline messages for anonymous accounts that nobody cares at all.

An anonymous account is a one-use and destroy account.

> offline messages are lost when user refreshes the page or closes the tab.

Ok, in that scenario from a user perspective, it would be desirable that the chat history is preserved at least some minutes, but not indefinitely. I wonder (like you) how to achieve that.

**Reply by licaon-kter** [reply]
ref: https://github.com/processone/ejabberd/issues/3737
